src/commands.rs

```diff
@@ -4071,9 +4073,85 @@ fn with_remote_callbacks<T>(ui: &mut Ui, f: impl FnOnce(git::RemoteCallbacks<'_>
         .map(|x| x as &mut dyn FnMut(&git::Progress));
     let mut get_ssh_key = get_ssh_key; // Coerce to unit fn type
     callbacks.get_ssh_key = Some(&mut get_ssh_key);
+    let mut get_pw = |url: &str, _username: &str| {
+        pinentry_get_pw(url).or_else(|| terminal_get_pw(&mut *ui.lock().unwrap(), url))
+    };
+    callbacks.get_password = Some(&mut get_pw);
+    let mut get_user_pw = |url: &str| {
+        let ui = &mut *ui.lock().unwrap();
+        Some((terminal_get_username(ui, url)?, terminal_get_pw(ui, url)?))
+    };
+    callbacks.get_username_password = Some(&mut get_user_pw);
     f(callbacks)
 }
 
+fn terminal_get_username(ui: &mut Ui, url: &str) -> Option<String> {
+    ui.prompt(&format!("Username for {}", url)).ok()
+}
+
+fn terminal_get_pw(ui: &mut Ui, url: &str) -> Option<String> {
+    ui.prompt_password(&format!("Password for {}: ", url)).ok()
+}
+
+fn pinentry_get_pw(url: &str) -> Option<String> {
```

Comment:
I've never used either `pinentry` or `rpassword`. I suppose the reason we attempt to use `pinentry` first is that it gives a better user experience than `rpassword` does, but what's the difference? Is it that it has some customization of the prompt that users may be used to (I'm just guessing)? Or better security?

Reply:
`pinentry` is an external binary that depending on configuration/environment may pop up a GUI using an appropriate toolkit, open an ncurses dialog, or display a traditional prompt. `rpassword` can only display a traditional prompt, but does not rely on an external binary, making it a good fallback.

Reply:
Ah, it can even display a GUI! I had not considered that. Thanks for explaining.